compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,505 @@
+use crate::assert::expr_if_not;
+use rustc_ast::{
+    ptr::P,
+    token::{self, Delimiter},
+    tokenstream::{DelimSpan, TokenStream, TokenTree},
+    Block, BorrowKind, Expr, ExprKind, ItemKind, LocalKind, MacArgs, MacCall, MacDelimiter,
+    Mutability, Path, PathSegment, Stmt, StmtKind, StructRest, UseTree, UseTreeKind, DUMMY_NODE_ID,
+};
+use rustc_ast_pretty::pprust;
+use rustc_expand::base::ExtCtxt;
+use rustc_span::{
+    symbol::{sym, Ident, Symbol},
+    Span, DUMMY_SP,
+};
+
+pub(super) struct Context<'cx, 'a> {
+    span: Span,
+    captures: Vec<Capture>,
+    cx: &'cx ExtCtxt<'a>,
+}
+
+impl<'cx, 'a> Context<'cx, 'a> {
+    pub(super) fn new(cx: &'cx ExtCtxt<'a>, span: Span) -> Self {
+        Self { cx, captures: Vec::new(), span }
+    }
+
+    /// Builds the whole `assert!` expression.
+    ///
+    /// {
+    ///    use ::core::asserting::{ ... };
+    ///
+    ///    let mut __capture0 = Capture::Unevaluated;
+    ///    ...
+    ///    ...
+    ///    ...
+    ///
+    ///    if !{
+    ///       ...
+    ///       ...
+    ///       ...
+    ///    } {
+    ///        panic!(
+    ///            "Assertion failed: ... \n With expansion: ...",
+    ///            (&__capture0).printer_kind().new(" ... ", __capture0),
+    ///            ...
+    ///            ...
+    ///            ...
+    ///        );
+    ///    }
+    /// }
+    pub(super) fn build(mut self, mut cond_expr: P<Expr>, panic_path: Path) -> P<Expr> {
+        let expr_str = escape_to_fmt(&pprust::expr_to_string(&cond_expr));
+        let mut fmt_str = String::new();
+        self.manage_cond_expr(&mut cond_expr, &mut fmt_str);
+        let initial_imports = self.build_initial_imports();
+        let panic = self.build_panic(&expr_str, &fmt_str, panic_path);
+
+        let Self { captures, cx, .. } = self;
+
+        let mut stmts = vec![initial_imports];
+        stmts.extend(captures.into_iter().map(|c| c.decl));
+        stmts.push(cx.stmt_expr(expr_if_not(cx, DUMMY_SP, cond_expr, panic, None)));
+        cx.expr_block(cx.block(DUMMY_SP, stmts))
+    }
+
+    /// Initial **trait** imports
+    ///
+    /// use ::core::asserting::{ ... };
```

Comment:
The suggestion to use full paths instead of imports applies to this too.

Reply:
Hum... I guess this is not possible because of the autoref-specialization stuff.

Two different traits have the save method name depending on whether `T` implements or does not implement some trait.

```rust
pub trait TryCaptureCopy<T> {
    fn try_capture(&self, to: &mut Capture<T>);
}

impl<T> TryCaptureCopy<T> for Wrapper<&T>
where
    T: Copy,
{
    #[inline]
    fn try_capture(&self, to: &mut Capture<T>) {
        *to = Capture::Value(*self.0);
    }
}

pub trait TryCaptureGeneric<T> {
    fn try_capture(&self, to: &mut Capture<T>);
}

impl<T> TryCaptureGeneric<T> for &Wrapper<&T> {
    #[inline]
    fn try_capture(&self, to: &mut Capture<T>) {
        *to = Capture::NoCopy;
    }
}

pub enum Capture<T> {
    NoCopy,
    Unevaluated,
    Value(T),
}

pub struct Wrapper<T>(pub T);

fn main() {
    let mut a = Capture::Unevaluated;
    let mut b = Capture::Unevaluated;
    (&Wrapper(&1i32)).try_capture(&mut a);
    (&Wrapper(&vec![1i32])).try_capture(&mut b);
    assert!(matches!(a, Capture::<i32>::Value(1)));
    assert!(matches!(b, Capture::<Vec<i32>>::NoCopy));
}
```

So something like `core::asserting::TryCaptureCopy::try_capture( ... )` can't be used because `T` is not known.